motoko/minimal-counter-dapp/README.md

```diff
@@ -0,0 +1,200 @@
+
+```html
+<!doctype html>
+<html lang="en">
+    <head>
+        <meta charset="UTF-8">
+        <meta name="viewport" content="width=device-width">
+        <title>hack</title>
+        <base href="/">
+
+        <link type="text/css" rel="stylesheet" href="main.css" />
+    </head>
+    <body>
+        <img src="logo.png" alt="DFINITY logo" />
+        <section>
+            <button id="clickMeBtn">Click Me!</button>
+        </section>
+        <section id="counter"></section>
+    </body>
+</html>
+```
+
+#### Javascript
+Two eventlisteners are added to the JavaScript file, `src/minimal_dapp/_assets/index.js`, the existing JavaScript file from the default project. One eventlistener is for detecting button clicks, and it's calling the `count()` function in the backend, and an eventlistener for page load is added to get the initial value of the counter with `getCount()`. The backend functions are imported through the Candid interface.
+
+```javascript
+import { minimaldapp } from "../../declarations/minimal_dapp";
+
+document.addEventListener('DOMContentLoaded', async function () {
+  const counter = await minimaldapp.getCount();
+  document.getElementById("counter").innerText = "Counter: " + counter;
+})
+
+document.getElementById("clickMeBtn").addEventListener("click", async () => {
+  const counter = await minimaldapp.count();
+  document.getElementById("counter").innerText = "Counter: " + counter;
+});
+```
+
+
+## Deployment
+The local network is started by running this command:
+
+```bash
+$ dfx start --background
+```
+
+When the local network is up and running, run this command to deploy the canisters:
+
+```bash
+$ dfx deploy
+```
+
+
+## Testing
+The functionality of this example dapp can be tested both in the frontend and in the backend. Before the example dapp can be tested, it must be deployed (locally) like described in the above Deployment section. 
+
+### Test the Frontend
+The URL for the frontend is depending on the canister ID, the local canister IDs can be found in `.dfx/local/canister_ids.json`. When deployed, the URL will look like this:
+
+**https://<ui_canister_id>.localhost:8000**
+
+### Test the backend
+There are two ways of testing the backend. One way is by making command line requests using DFX, and the other way is to use the Candid UI.
+
+#### DFX
+DFX has a subset of commands for canister operations, and one of them enables calling the public functions added to the `main.mo` file in the previous step. In the following examples the initial value is 0. `count` will increment value and return 1, `getCount` will return the current value and `reset` will set the value to 0.
+
+Command usage: `dfx canister call &#60project&#62  &#60function&#62`
+
+```bash
+$ dfx canister call minimal_dapp count
+(1 : Nat)
+```
+
+```bash
+$ dfx canister call minimal_dapp getCount
```

Comment:
```suggestion
$ dfx canister call minimal_dapp getCount
```
I don't know if this is still the case, but in the past you had to add `--query` to get the faster call semantics. Otherwise I think the query is executed as a slower ingress message, with consensus (but no lasting state change). I maybe be wrong though.

Reply:
I think I'll leave it out for simplicity, but you have a good point.

Reply:
`dfx canister call` now by default makes a query call for queries, or an update call for updates.

```
$ time dfx canister call minimal_dapp count   
(9 : nat)
dfx canister call minimal_dapp count  0.04s user 0.01s system 3% cpu 1.502 total
$ time dfx canister call minimal_dapp count
(10 : nat)
dfx canister call minimal_dapp count  0.04s user 0.01s system 3% cpu 1.534 total
$ time dfx canister call minimal_dapp getCount
(10 : nat)
dfx canister call minimal_dapp getCount  0.01s user 0.01s system 59% cpu 0.038 total
$ time dfx canister call minimal_dapp getCount
(10 : nat)
dfx canister call minimal_dapp getCount  0.01s user 0.01s system 63% cpu 0.037 total
```